src/iterators.jl

```diff
@@ -1,145 +1,190 @@
+    if cc.flags.nodes
+        cellnodes!(cc.nodes, cc.grid, i)
+    end
+    if cc.flags.coords
+        cellcoords!(cc.coords, cc.grid, i)
+    end
+    if cc.dh !== nothing && cc.flags.dofs
+        @assert cc.dh isa DofHandler
+        celldofs!(cc.dofs, cc.dh, i)
     end
+    return cc
 end
-Base.length(ci::CellIterator)  = ci.cellset === nothing ? length(ci.grid.cells) : length(ci.cellset)
-
-Base.IteratorSize(::Type{T})   where {T<:CellIterator} = Base.HasLength() # this is default in Base
-Base.IteratorEltype(::Type{T}) where {T<:CellIterator} = Base.HasEltype() # this is default in Base
-Base.eltype(::Type{T})         where {T<:CellIterator} = T
-
-# utility
-@inline getnodes(ci::CellIterator) = ci.nodes
-@inline getcoordinates(ci::CellIterator) = ci.coords
-@inline nfaces(ci::CellIterator) = nfaces(eltype(ci.grid.cells))
-@inline onboundary(ci::CellIterator, face::Int) = ci.grid.boundary_matrix[face, ci.current_cellid[]]
-@inline cellid(ci::CellIterator) = ci.current_cellid[]
-@inline celldofs!(v::Vector, ci::CellIterator) = celldofs!(v, ci.dh, ci.current_cellid[])
-@inline celldofs(ci::CellIterator) = ci.celldofs
-
-function reinit!(ci::CellIterator{dim,C}, i::Int) where {dim,C}
-    ci.current_cellid[] = ci.cellset === nothing ? i : ci.cellset[i]
-
-    if ci.flags.nodes
-        if ci.dh !== nothing && ci.dh isa MixedDofHandler
-            cellnodes!(ci.nodes, ci.dh, ci.current_cellid[])
-        else
-            cellnodes!(ci.nodes, ci.grid, ci.current_cellid[])
-        end
+function reinit!(cc::CellCache{<:Any,<:AbstractGrid,<:MixedDofHandler}, i::Int)
+    @assert cc.dh isa MixedDofHandler
+    cc.cellid[] = i
+    if cc.flags.nodes
+        resize!(cc.nodes, nnodes_per_cell(cc.dh, i))
+        cellnodes!(cc.nodes, cc.dh, i)
     end
-    if ci.flags.coords
-        if ci.dh !== nothing && ci.dh isa MixedDofHandler
-            cellcoords!(ci.coords, ci.dh, ci.current_cellid[])
-        else
-            cellcoords!(ci.coords, ci.grid, ci.current_cellid[])
-        end
+    if cc.flags.coords
+        resize!(cc.coords, nnodes_per_cell(cc.dh, i))
+        cellcoords!(cc.coords, cc.dh, i)
     end
-
-    if ci.dh !== nothing && ci.flags.celldofs # update celldofs
-        celldofs!(ci.celldofs, ci.dh, ci.current_cellid[])
+    if cc.flags.dofs
+        resize!(cc.dofs, ndofs_per_cell(cc.dh, i))
+        celldofs!(cc.dofs, cc.dh, i)
     end
-    return ci
+    return cc
 end
 
-function check_compatible_geointerpolation(cv::Union{CellValues, FaceValues}, ci::CellIterator)
-    if length(getnodes(ci)) != getngeobasefunctions(cv)
-        msg = """The given CellValues and CellIterator are incompatiblet.
-        Likely an appropriate geometry interpolate must be passed when constructing the CellValues.
-        See also issue #265: https://github.com/Ferrite-FEM/Ferrite.jl/issues/265"""
-        throw(ArgumentError(msg))
-    end
+# reinit! FEValues with CellCache
+reinit!(cv::CellValues, cc::CellCache) = reinit!(cv, cc.coords)
```

Comment:
Personally i never liked this method...I think it hides to much information to new users.
Using `reinit!(cv, getcoordinates(cc))` is clearer i think.

Reply:
Sure, but could be in a cleanup PR, this is just replicating the functionality we already have :)